activesupport/lib/active_support/core_ext/module/delegation.rb

```diff
@@ -197,9 +197,34 @@ def delegate(*methods, to: nil, prefix: nil, allow_nil: nil, private: nil)
       method_name = prefix ? "#{method_prefix}#{method}" : method
       method_names << method_name.to_sym
 
+      begin
+        arity = \
+          if to.is_a?(Class)
+            to.method(method).arity
+          elsif receiver == "self.class"
+            method(method).arity
+          end
+      rescue NameError => e
+        raise NameError, "Definition of #{e.receiver.name}##{method} couldn't be found. Perhaps you're trying to delegate before defining that method?"
```

Comment:
Alternatively we could just fallback to the old way if the method we're supposed to delegate to doesn't exist yet.

Because I fear this may cause a lot of backward compatibility issues, circular dependencies etc.

Reply:
Yeah, I think assuming the eventual delegation target will have the same arity as the one that exists at delegation-definition time is reasonable, but forcing sequencing is a needlessly strong breaking change.

Going for a faster option when we can, but silently falling back to a reasonable-but-slightly-slower behaviour is a good balance IMO. It's nicely faster in a microbenchmark, but I don't think it's significant enough to force an error + code change onto callers so they can benefit.

Reply:
I also wonder if it would make sense to allow something like:

```ruby
delegate :role, to: :user, as: User
```

As an indication that we're delegating to `User#role` and that we should use its arity for optimized delegation.

Reply:
> Alternatively we could just fallback to the old way if the method we're supposed to delegate to doesn't exist yet.

Yes, I firstly implemented like that, then changed my mind, because I realized that if it silently falls back, nobody can notice the fact that they are choosing the slower path... But I can understand the opinions that the actual performance gain might not worth introducing a new incompatibility. I'll revert back to my initial implementation then.

Reply:
OK, just removed the incompatibility. Now it falls back to the original behavior when target class was not found.

Reply:
> `delegate :role, to: :user, as: User`

Yes! That's exactly what I have in my mind, and that's what I meant by
> I will work on further extension on delegate to take a new option, probably arity or class once this PR is merged.

in the "Future Plan" section in this PR body. I'm actually trying that already, and to me it feels better than passing in `:arity` value per each method. I chose the name `class:` in my working branch, but I can rename it anytime.

Reply:
> Yes! That's exactly what I have in my mind, and that's what I meant by

Oh sorry, I missed it.

>  I chose the name class: in my working branch, but I can rename it anytime.

Hum. I guess `class:` would work but it seem a be vague to me. I kinda like `:as` as it implies casting / typing. But I don't feel particularly strongly about it.